### DISCUSSION TITLE
Authorization services and adapter deprecation

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#11681] [Adapter deprecation]
There is wide-spread support for OIDC for various languages and frameworks today. However, this does not apply to authorization services. As such we need to provide at least a basic Java client for authorization services that can be integrated into different Java applications through for example a servlet filter.

**Comment by naveedhassan** [comment]
So if I understand correctly there is no way to leverage the centralized authorization services like policy enforcer right now as the adapters gets deprecated?

**Reply by pedroigor** [reply]
Right now, only the Elytron-based adapter is available and you can still configure a policy enforcer for your applications.

But once we remove the OIDC adapter completely it won't be available anymore and we need to provide something like what Stian is suggesting.

**Reply by hean0224** [reply]
Looks like Quarkus implementation supports things like http-method-as-scope and policyenforcerconfig.  I might take a look at transitioning my project to Quarkus from Spring. https://quarkus.io/guides/security-keycloak-authorization

**Comment by stianst** [comment]
@pedroigor FIY

**Comment by paulceli** [comment]
Was wondering the same as @naveedhassan.  Doesn't feel right to start an implementation on a deprecated adapter. 
Is there something available in spring security libs to leverage the keycloak auhtorizations ? any advice or roadmap ?  thx

**Comment by pedroigor** [comment]
@paulceli You shouldn't find anything from Spring or any other stack that integrates to Keycloak Authorization Services. I won't be surprised though if there is a project, even if maintained by an individual, available from GitHub.

In any case, we should be providing some pure Java-based API that can integrate easily on different stacks.

The policy enforcer available from the adapter is self-contained and can be extracted as a generic library.

**Reply by paulceli** [reply]
Are you talking about keycloak-authz-client ? 
It seems unrelated to spring. Why not recommend to continue using just that part ?

**Reply by pedroigor** [reply]
That library provides the "connection" bits, and an API to interact with the server to obtain permissions and use the Protection API.

But the PEP also has a layer on top of that with the logic to actually process HTTP requests and grant/deny access accordingly.

**Comment by GretUp** [comment]
I seriously hope that the decision not to support policy enforcer any more will be reconsidered.

**Reply by pedroigor** [reply]
We can definitely provide some generic library for Java, something easy to integrate using a filter, interceptor, etc.

**Reply by GretUp** [reply]
@pedroigor is there a first version of that generic java library available? What is the timeline here?

**Comment by tbroadbent22** [comment]
Has anyone had much luck migrating to pac4j, particularly regarding dropwizard?
See https://www.pac4j.org/

**Comment by Vespasian** [comment]
Was there a final decision made on whether such a library will be provided?

**Comment by minusdavid** [comment]
I've been wondering how to implement Keycloak authorization services and I only noticed examples using client adapters...

Does that mean there's no reference implementation aside from the deprecated client adapters?

Doesn't Red Hat use RHSSO/Keycloak for AuthN and AuthZ for redhat.com?

**Comment by minusdavid** [comment]
At the moment, if I wanted to use Keycloak authorization services, I'm guessing I'd need to write my own client?

If I understand UMA 2.0...
1. Human request party interacts with a client (e.g. Web App)
2. The client tries to access resource server (e.g. Web App 2 or sub-component of Web App) without a RPT
3. The resource server asks the authorization server (Keycloak) for a permission ticket
4. The resource server returns the permission ticket and authorization server's location
5. The client requests RPT from the authorization server using the permission ticket

I notice there is an access bearer token at https://docs.kantarainitiative.org/uma/ed/uma-core-2.0-21.html#rfc.section.3.6.1 but it's not clear if that access token belongs to the client or the human requesting party. I assume that it's a OIDC access token obtained by the client for the human requesting party before any of this UMA 2.0 flow...

6. In theory, we have permission, so the authorization server (Keycloak) returns a RPT to the client
7. The client attempts to access resource server again using the RPT. 
8. The resource server validates the RPT permissions if the RPT is a JWT (or else queries the authorization server if it's just an arbitrary token)
9. If RPT contains valid permissions, the resource server serves the resource to the client (e.g. Web App)

--

This process seems useful for infrequent constrained requests, but inefficient if you're frequently accessing different resources with different actions/scopes. 

Although if the authorization server supports RPT upgrading, then you might build up a RPT with the necessary permissions quickly. I suppose it depends on how the resource server structures its resources and the authorization policies...

**Reply by minusdavid** [reply]
Actually, upon reflection, it looks like the client uses its own client id and client secret for Step 5...

And since it's representing a human requesting party, I think it needs to include an OIDC ID token via the "claim_token" and "claim_token_format" parameters...

Without those parameters, I imagine the authorization server would need interaction from the human requesting party to provide claim information. (Alternatively, maybe it use a client redirection to pickup the Keycloak session cookie to fetch these details for itself?)

**Reply by noah-hein** [reply]
I'm in the same boat as you. I've been really trying to use Authz for Spring via my keycloak server but stuck trying to create an implementation with the bare bones OAuth2 client.

I was thinking about creating an @PreAuthorize annotation method to check if the user has access to the resource scope. My idea was to put this annotation on different endpoints specifying the scope / resource pair. The method would first check if the current token contains the required authorization. If it does not, then a token request would be made with the scope / resource for an RPT. 

The new token should return with the provided info. If it doesn't, then the client doesn't have access. If the new RPT contains the valid authorization info, then replace the old auth token (to try and reduce required requests)

It's really unfortunate that the only solution to policy enforcement rn is using Quarkus. I might check it out.

**Reply by Vespasian** [reply]
Looking at the code in the Quarkus plugin I think this is pretty much what they do.

Check whether the token has the resource permissions and request them from keycloak if it doesn't.

**Comment by minusdavid** [comment]
One last thing...

I like how https://github.com/keycloak/keycloak-quickstarts/tree/latest/app-authz-springboot says "the AuthorizationContext encapsulates all permissions granted by a Keycloak server and provides methods to check these permissions". 

That would be really useful.

**Comment by kkcmadhu** [comment]
woudl it impact token exchange and admin-fine-grained-authz ?
because these internally creates policies etc i guess.

**Reply by pedroigor** [reply]
Not really, as the internals are going to be kept or even replaced. The deprecation here is related to the capabilities you get from adapters.

**Comment by christophenoel** [comment]
Do you confirm that there is still currently no alternative to KeyCloak adapter for enforcing keycloak policies ?

**Reply by pedroigor** [reply]
@christophenoel There is, but only for Java.

The policy enforcer code [1] was extracted from our adapter code base into a separate dependency.

With that, Java applications are going to be able to use the policy enforcer on any Java-stack (not only those supported by our adapters).

We are working on documentation and how to use it. The documentation should be initially based on how to enable it to Jakarta applications. For instance, using a servlet filter to enforce policies.

[1] https://github.com/keycloak/keycloak/tree/main/authz/policy-enforcer

**Reply by christophenoel** [reply]
Yes, I was very happy this morning when I got these news. ! Thanks.

**Comment by winterrobert** [comment]
In a similar manner to @minusdavid we have been relying on the node.js client Keycloak.enforcer() method to control access to different resources based on the URL (e.g. user X should only have access to /resourceX) but I'm unsure how to proceed at this point.

I'm not super keen on implementing and maintaining this UMA logic ourselves. Even though the [openid-client + passport combo](https://dev.to/austincunningham/keycloak-express-openid-client-3mmg) could be a replacement for the core authentication I haven't been able to find any replacement out there for the [User Managed Access part](https://www.keycloak.org/docs/latest/authorization_services/#_service_uma_authorization_process).

What are other ways of handling this - would redesigning to some simpler authorization approach be better? E.g. handling access control by group memberships or something else?

**Comment by trodix** [comment]
I have some good news! :1st_place_medal: :100: 

@pedroigor did a great job to decouple the code as you can see on the PR: https://github.com/keycloak/keycloak/issues/17353

The PR had been merged into the keycloak main branch and labeled as 22.0 milestone that should be released in March / April 2023 => https://www.keycloak.org/2022/03/releases.html#:~:text=Keycloak%2022%20%2D%20March%2FApril%202023

I'm waiting for the official announcement! Thank you so much :heart:

**Reply by minusdavid** [reply]
Sounds interesting although it also sounds like a Java only solution? We use Keycloak with applications implemented in all kinds of different languages, so sounds like we might still be out of luck.

**Reply by pedroigor** [reply]
Yeah, unfortunately, we only have the capacity to support/maintain it for Java.

**Reply by minusdavid** [reply]
Fair enough! Perhaps you can help me understand the Java implementation better.

I'm looking at https://github.com/keycloak/keycloak/blob/main/authz/policy-enforcer/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java but I'm not a Java expert. I'm assuming that an implementer would create the PolicyEnforcer object and then pass the HTTP request to the enforce() method and it would take it from there? 

At a high level, the ideal scenarios seem to be it locally gets the stored access token for the authenticated user, checks locally if the access token's permissions match the requested permissions. If it is, return authorized. If not, it depends on whether or not it's UMA protected. 

Not UMA protected: use the stored access token (obtained via OIDC) to fetch a new access token (ie RPT) for the path specified from Keycloak. If an access token (ie RPT) is returned, update the permissions for the stored access token for the authenticated user, check locally if the access token's permissions match the requested permissions, and return authorized. 

If UMA protected: the library returns a 401 with a UMA ticket. And this is where I get a bit confused. If you're an API consumer, it's not a big deal. You can use the ticket to obtain a RPT, and repeat the process. But if you're a web application... I suppose you could code the same thing. It just seems like more steps than the "non-UMA" route. Also seems like it would be harder to embed your PEP as a filter/middleware/etc although perhaps not.

Anyway, happy to pursue this more on Discourse or email or whatever. I just want to make sure that I'm on the right track 😅

**Reply by harshdesai991** [reply]
@pedroigor Thanks for extracting policy enforcers for Java Lib. Could you please let us know when is the release planned for it?

**Reply by pedroigor** [reply]
@minusdavid That is it, in general. I wouldn't worry about UMA in the first iteration but focus on the core functionality. UMA can be enabled later as it is more about leveraging the core plus how to properly respond to clients.

**Reply by pedroigor** [reply]
@harshdesai991 It is already upstream and it is targeted for Keycloak 22. @stianst can give you more details about the release plan.

**Reply by harshdesai991** [reply]
@stianst I know Keycloak 22 was planned for March / April 2023. Do you have any latest updates on the release?
@pedroigor Is the documentation on using the new Policy Enforcer library available now?

**Reply by pedroigor** [reply]
@harshdesai991 Not yet available. But I'm working on some quickstarts, you should see commits arriving there soon.

I'm not 100% sure if we need specific documentation for the policy enforcer but focus on code examples from our quickstart repository. Wdyt?

**Reply by harshdesai991** [reply]
@pedroigor yeah code examples would be great  as well 👍

**Reply by harshdesai991** [reply]
@pedroigor To get started before releasing of 22 version, can I use the nightly release of Keycloak? Will it have relevant changes related to your policy enforcer change?

**Reply by pedroigor** [reply]
Yes, you can. A quickstart is being developed here https://github.com/keycloak/keycloak-quickstarts/pull/396.

**Reply by harshdesai991** [reply]
@pedroigor I am trying to integrate keycloak-policy-enforcer with Spring Boot 3 application. Looking at your above PR I created oidc.json and policy-enforcer.json in resources folder, than I am trying to get AuthorizationContext in OncePerRequestFilter like below but I am getting the AuthorizationContext as **NULL** Could you please point me in the right direction to integrate this with Spring Boot application :

```
@Component
public class KeycloakAuthzFilter extends OncePerRequestFilter {
    private static final Logger LOGGER = LoggerFactory.getLogger(KeycloakAuthzFilter.class);

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        AuthorizationContext authorizationContext = (AuthorizationContext) request.getAttribute(AuthorizationContext.class.getName());
        if (authorizationContext == null) {
            LOGGER.error("No authorization context found in request.");
            return;
        }
        for (Permission permission : authorizationContext.getPermissions()) {
            LOGGER.info("Resource: {}", permission.getResourceName());
            LOGGER.info("ID: {}", permission.getResourceId());
            LOGGER.info("Scopes: {}", permission.getScopes());
            LOGGER.info("Claims: {}", permission.getClaims());
        }
        filterChain.doFilter(request, response);
    }
}
```

**Comment by zak905** [comment]
Will the adapters be removed in subsequent releases ? meaning they will be no artifacts ?

**Reply by trodix** [reply]
Yes, See at the bottom of this page: https://www.keycloak.org/2022/03/releases.html

**Reply by zak905** [reply]
thanks for the info. This is sad though

**Comment by stianst** [comment]
For updated plans around adapters please see https://www.keycloak.org/2023/03/adapter-deprecation-update.html. In terms of Authorization Services we are planning on providing a generic Java library that can be leveraged in combination with other OIDC/OAuth libraries, we are also planning to provide support for Node.js and Authorization Service regardless of the future plans around that adapter as well.

**Reply by minusdavid** [reply]
I had to re-read all the comments and a bit of https://www.keycloak.org/docs/latest/authorization_services/ to try to wrap my head around everything again, but it sounds interesting! 

@stianst : it makes sense "leaving the integration for various programming languages and frameworks to the relevant communities". I'm not opposed to writing my own library (I've done it for OIDC in the past using the OIDC spec to guide me), but I'm still trying to wrap my head around UMA and Keycloak's Authorization Services. 

Hopefully, when the Java and Node.js libraries are fully ready, I could use them as a basis for a library in another programming language.

**Reply by winterrobert** [reply]
Great that you're keeping the node.js adapter for Authorization around, thanks for that.

**Reply by pedroigor** [reply]
@minusdavid I wouldn't worry about enabling UMA, at least in the first iteration/version.

The PEP implementation is basically about querying the token endpoint for an access token with permissions inside (granted by the authz service) and mapping those permissions to paths in your application. If the permission for a given path is there, move forward with the request.

Devil is in the details but that is the idea.

**Reply by minusdavid** [reply]
The devil is totally in the details haha, but I appreciate the advice! I have a bunch of questions but I figure I should try having a few goes at it first.

What's the best place to ask follow-up questions about authorization services?

**Reply by pedroigor** [reply]
A separate discussion should be fine.
